password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,222 @@
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+{-|
+Module      : Data.Password.Valid
+Copyright   : (c) Dennis Gosnell, 2019; Felix Paulusma, 2020; Hiroto Shioi, 2020
+License     : BSD-style (see LICENSE file)
+Maintainer  : cdep.illabout@gmail.com
+Stability   : experimental
+Portability : POSIX
+
+= validate
+
+It is common for passwords to have a set of requirements.
+(e.g. Password must contain at least eight characters that consist of alphabetic
+characters combined with numbers or special characters.)
+
+Validate module provides set of functions which enables you to validate them.
+
+-}
+
+module Data.Password.Validate
+  ( -- * Data type
+    PasswordPolicy (..),
+    InvalidReason (..),
+    CharacterCategory(..),
+    -- * Default
+    defaultPasswordPolicy,
+    -- * Functions
+    isValidPassword,
+    validatePassword,
+    -- * Utility
+    isValidPasswordPolicy,
+    defaultCharSet,
+  ) where
+
+import Data.Char (isDigit, isLower, isUpper)
+import Data.Maybe (catMaybes, fromMaybe)
+import Data.Password.Internal (Password (..))
+#if! MIN_VERSION_base(4,13,0)
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy = PasswordPolicy
+    { minimumLength  :: !Int
+    -- ^ Required password minimum length
+    , maximumLength  :: !Int
+    -- ^ Required password maximum length
+    , uppercaseChars :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , lowercaseChars :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , specialChars   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , digitChars     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    , charSet        :: !Text
```

Comment:
After thinking about this for a while, it might be a good idea to change the `charSet` field from `Text` to `Char -> Bool`? (maybe rename then), because if someone wants to allow all Chinese/Japanese characters and emojis, that's pretty much all of unicode, so that would mean they'd need an unnecessarily large list.

Reply:
Yes, that is very good point. The function needs to be wrapped with the newtype so that I can derive `Show` instance of it.
e.g.

```haskell
newtype CharSetRule = { getCharSetRule :: Char -> Bool }

instance Show CharSetRule where
   show _ = "Character set rule"
```

Reply:
@Vlix Yes, that's a good point. I also think we need to wrap the function with newtype so that it's much easier to derive typeclass instances.  

e.g.

```haskell
newtype CharacterSetRule = CharacterSetRule { getCharacterSetRule :: Char -> Bool }

instance Show CharacterSetRule where
   show _  = "Character set rule"
```

If you agree, what do you think the name should be?

Reply:
Or what we can do it remove the field, and add an extra argument to existing functions.

So instead of

```haskell
isValidPassword :: PasswordPolicy -> Password -> Bool
```

we can do

```haskell
isValidPassword :: PasswordPolicy -> CharSetRule -> Password -> Bool
```

Reply:
Taking it out of the `PasswordPolicy` might be a good idea, but keeping the newtype would make sense too.

Adding a `Semigroup/Monoid` instance could make it easy(/easier?) to compose rules. (kind of like how the [`retry`](https://hackage.haskell.org/package/retry-0.8.1.2/docs/Control-Retry.html#t:RetryPolicyM) package does it) Though not too sure we'd need composition. Most people will go with the default and other than maybe excluding the special characters, I don't see much convenience rules we could make that people would want to compose.

Example:
```hs
-- alphanumeric = [0-9a-zA-Z]
-- latin1Letters = [À-ÖØ-öø-ÿ]
myCharSet = alphanumeric <> latin1Letters
```

Reply:
Yeah, adding `Semigroup/Monoid` is possible, but I think most people would default value.